lib/livebook/application.ex

```diff
@@ -180,33 +182,84 @@ defmodule Livebook.Application do
     end
   end
 
-  if Mix.target() == :app do
+  @app? Mix.target() == :app
+
+  if @app? do
     defp app_specs, do: [LivebookApp]
   else
     defp app_specs, do: []
   end
 
+  # In order to provide good first experience with the desktop app,
+  # in case the endpoint or iframe port is taken, we automatically
+  # fallback to a random port.
+
+  if @app? do
+    defp endpoint_childspec(opts) do
+      %{start: start} = childspec = LivebookWeb.Endpoint.child_spec(opts)
+      %{childspec | start: {__MODULE__, :endpoint_start, [start]}}
+    end
+  else
+    defp endpoint_childspec(opts), do: LivebookWeb.Endpoint.child_spec(opts)
+  end
+
+  @doc false
+  def endpoint_start({mod, fun, args}) do
+    with {:error,
+          {:shutdown,
+           {:failed_to_start_child, {LivebookWeb.Endpoint, :http},
+            {:shutdown, {:failed_to_start_child, :listener, :eaddrinuse}}}}} <-
+           apply(mod, fun, args) do
+      config = Application.get_env(:livebook, LivebookWeb.Endpoint)
+      config = put_in(config[:http][:port], 0)
+      Application.put_env(:livebook, LivebookWeb.Endpoint, config, persistent: true)
+      Logger.warning("Starting server using a random port")
+      endpoint_start({mod, fun, args})
+    end
+  end
+
   defp iframe_server_specs() do
     server? = Phoenix.Endpoint.server?(:livebook, LivebookWeb.Endpoint)
     port = Livebook.Config.iframe_port()
 
     if server? do
       http = Application.fetch_env!(:livebook, LivebookWeb.Endpoint)[:http]
 
-      iframe_opts =
+      opts =
         [
           scheme: :http,
           plug: LivebookWeb.IframeEndpoint,
           port: port,
           thousand_island_options: [supervisor_options: [name: LivebookWeb.IframeEndpoint]]
         ] ++ Keyword.take(http, [:ip])
 
-      [{Bandit, iframe_opts}]
+      [iframe_endpoint_childspec(opts)]
     else
       []
     end
   end
 
+  if @app? do
+    defp iframe_endpoint_childspec(opts) do
+      %{start: start} = childspec = Bandit.child_spec(opts)
+      %{childspec | start: {__MODULE__, :iframe_endpoint_start, [start]}}
+    end
+  else
+    defp iframe_endpoint_childspec(opts), do: Bandit.child_spec(opts)
+  end
+
+  @doc false
+  def iframe_endpoint_start({mod, fun, args}) do
+    with {:error, {:shutdown, {:failed_to_start_child, :listener, :eaddrinuse}}} <-
+           apply(mod, fun, args) do
+      Application.put_env(:livebook, :iframe_port, 0, persistent: true)
+      [opts] = args
+      args = [Keyword.replace(opts, :port, 0)]
+      Logger.warning("Starting iframe server using a random port")
+      iframe_endpoint_start({mod, fun, args})
+    end
+  end
```

Comment:
if `args` is always `[opts]`, another idea is to be super explicit about that which buy us one less line (the `[opts] = args` match). 😅  Fine as is obviously!

```suggestion
  def iframe_endpoint_start({mod, fun, [opts]}) do
    with {:error, {:shutdown, {:failed_to_start_child, :listener, :eaddrinuse}}} <-
           apply(mod, fun, [opts]) do
      Application.put_env(:livebook, :iframe_port, 0, persistent: true)
      opts = Keyword.replace(opts, :port, 0)
      Logger.warning("Starting iframe server using a random port")
      iframe_endpoint_start({mod, fun, [opts]})
    end
  end
```